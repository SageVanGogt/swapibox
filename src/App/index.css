.App {
  display: grid;
  grid-template-columns: 200px 1fr 400px;
  grid-template-rows: 300px 1fr;
  grid-template-areas: "a b c" "e e f";
  justify-content: stretch;
  align-items: stretch; }

.nav-container {
  grid-area: f; }

.beer-nav {
  height: 400px; }

.cantina-image {
  position: absolute;
  height: 100%;
  z-index: -1; }
