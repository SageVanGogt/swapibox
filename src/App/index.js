import React, { Component } from 'react';
import Button from './../Button/index';
import Film from './../Film/index';
import CardContainer from './../CardContainer/index';
import fetchApiData from './../api';
import { getRandomFilm } from './../helper';
import './index.css';

class App extends Component {
  constructor() {
    super();

    this.state = {
      currentRandomFilm: {},
      currentSectionData: [],
      allData: {vehicles: [], planets: [], people: [], favorites: [], favoriteCount: 0}
    };
  }

  async componentDidMount() {
    const film = await this.fetchFilm();
    const currentRandomFilm = await getRandomFilm(film);
    this.setState({
      currentRandomFilm
    })
  }

  fetchFilm = async () => {
    const url = 'https://swapi.co/api/films/'
    const response = await fetch(url);
    const film = await response.json();
    return film;
  }

  handleClickEvent = async (event) => {
    const { name } = event.target;
    if(this.state.allData[name].length || name === 'favorites') {
      this.setStateToExistingData(name);
      return;
    }
   
    const currentSectionData = await fetchApiData(name); 
    const allData = {...this.state.allData, [name]: currentSectionData}
    await this.setState({
      currentSectionData, 
      allData
    })
  }

  setStateToExistingData = (name) => {
    const currentSectionData = this.state.allData[name];

    this.setState({
      currentSectionData
    })
  }

  handleFavorite = (event) => {
    const { value } = event.target;
    const currentChosen = this.state.currentSectionData.find(data => data.name === value)
    const favorites = this.state.allData.favorites;
    const favoriteExists = favorites.find(data => data.name === value);
    //includes filter out, if it does spread it in
    //refactor this area
    currentChosen.favorited = !currentChosen.favorited;

    if(favoriteExists) {
      const favoriteIndex = favorites.indexOf(favoriteExists);
      favorites.splice(favoriteIndex, 1);
    } else {
      favorites.push(currentChosen)
    }
    const favoriteCount = favorites.length
    const allData = {...this.state.allData, favorites, favoriteCount}
    
    this.setState({
      allData
    })
  }

  render() {
    return (
      <div className="App">
        <img 
          src="/assets/cantina.jpg" 
          alt="background image of cantina" 
          className="cantina-image"
        />
        <div className="nav-container">
          <nav>
            <Button 
              section={'people'} 
              handleClickEvent={this.handleClickEvent}
            />
            <Button 
              section={'planets'} 
              handleClickEvent={this.handleClickEvent}
            />
            <Button 
              section={'vehicles'} 
              handleClickEvent={this.handleClickEvent}
            />
            <Button 
              section={'favorites'} 
              handleClickEvent={this.handleClickEvent} 
              favoriteCount={this.state.allData.favoriteCount}
            />
          </nav>
          <img 
            src="./assets/beerTap.png" 
            alt="beertap navigation" 
            className="beer-nav"
          />
        </div>
        { 
          this.state.currentRandomFilm && 
          <Film 
            currentRandomFilm={this.state.currentRandomFilm}
          />
        }
        {
          this.state.currentSectionData &&
          <CardContainer
            currentSectionData={this.state.currentSectionData}
            handleFavorite={this.handleFavorite}
          />
        }
      </div>
    );
  }
}

export default App;
