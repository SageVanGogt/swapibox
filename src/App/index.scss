
.App {
  display: grid;
  grid-template-columns: 200px 1fr 400px;
  grid-template-rows: 300px 1fr;
  grid-template-areas: 
  "a f f"
  "e e e";
  justify-content: stretch;
  align-items: stretch;
  // background-image: url('/assets/cantina.jpg');
  // background-size: 120vw;
  // background-repeat: no-repeat;
}

.nav-container {
  grid-area: f;
  margin-left: 500px;
  margin-top: 50px;
}

.beer-nav {
  height: 600px;
}

.cantina-image {
  position: absolute;
  height: 100%;
  z-index: -1;
}

nav {
  display: flex;
}
