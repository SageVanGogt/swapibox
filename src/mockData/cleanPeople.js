const cleanPeople = [ { 
name: 'Luke Skywalker',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'C-3PO',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'R2-D2',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'Darth Vader',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'Leia Organa',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'Owen Lars',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'Beru Whitesun lars',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'R5-D4',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'Biggs Darklighter',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false },
{ name: 'Obi-Wan Kenobi',
species:  undefined ,
homeworld: undefined,
homeworldPopulation: undefined,
favorited: false } ]

export default cleanPeople;