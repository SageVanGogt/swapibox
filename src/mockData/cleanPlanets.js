export default cleanPlanets = [ { name: 'Alderaan',
terrain: 'grasslands, mountains',
population: '2000000000',
climate: 'temperate',
residents: 
 [ 'https://swapi.co/api/people/5/',
   'https://swapi.co/api/people/68/',
   'https://swapi.co/api/people/81/' ],
favorited: false },
{ name: 'Yavin IV',
terrain: 'jungle, rainforests',
population: '1000',
climate: 'temperate, tropical',
residents: [],
favorited: false },
{ name: 'Hoth',
terrain: 'tundra, ice caves, mountain ranges',
population: 'unknown',
climate: 'frozen',
residents: [],
favorited: false },
{ name: 'Dagobah',
terrain: 'swamp, jungles',
population: 'unknown',
climate: 'murky',
residents: [],
favorited: false },
{ name: 'Bespin',
terrain: 'gas giant',
population: '6000000',
climate: 'temperate',
residents: [ 'https://swapi.co/api/people/26/' ],
favorited: false },
{ name: 'Endor',
terrain: 'forests, mountains, lakes',
population: '30000000',
climate: 'temperate',
residents: [ 'https://swapi.co/api/people/30/' ],
favorited: false },
{ name: 'Naboo',
terrain: 'grassy hills, swamps, forests, mountains',
population: '4500000000',
climate: 'temperate',
residents: 
 [ 'https://swapi.co/api/people/3/',
   'https://swapi.co/api/people/21/',
   'https://swapi.co/api/people/36/',
   'https://swapi.co/api/people/37/',
   'https://swapi.co/api/people/38/',
   'https://swapi.co/api/people/39/',
   'https://swapi.co/api/people/42/',
   'https://swapi.co/api/people/60/',
   'https://swapi.co/api/people/61/',
   'https://swapi.co/api/people/66/',
   'https://swapi.co/api/people/35/' ],
favorited: false },
{ name: 'Coruscant',
terrain: 'cityscape, mountains',
population: '1000000000000',
climate: 'temperate',
residents: 
 [ 'https://swapi.co/api/people/34/',
   'https://swapi.co/api/people/55/',
   'https://swapi.co/api/people/74/' ],
favorited: false },
{ name: 'Kamino',
terrain: 'ocean',
population: '1000000000',
climate: 'temperate',
residents: 
 [ 'https://swapi.co/api/people/22/',
   'https://swapi.co/api/people/72/',
   'https://swapi.co/api/people/73/' ],
favorited: false },
{ name: 'Geonosis',
terrain: 'rock, desert, mountain, barren',
population: '100000000000',
climate: 'temperate, arid',
residents: [ 'https://swapi.co/api/people/63/' ],
favorited: false } ]