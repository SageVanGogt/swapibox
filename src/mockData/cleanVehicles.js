export default cleanVehicles = [ { name: 'Sand Crawler',
model: 'Digger Crawler',
vehicleClass: 'wheeled',
passengerCount: '30',
favorited: false },
{ name: 'T-16 skyhopper',
model: 'T-16 skyhopper',
vehicleClass: 'repulsorcraft',
passengerCount: '1',
favorited: false },
{ name: 'X-34 landspeeder',
model: 'X-34 landspeeder',
vehicleClass: 'repulsorcraft',
passengerCount: '1',
favorited: false },
{ name: 'TIE/LN starfighter',
model: 'Twin Ion Engine/Ln Starfighter',
vehicleClass: 'starfighter',
passengerCount: '0',
favorited: false },
{ name: 'Snowspeeder',
model: 't-47 airspeeder',
vehicleClass: 'airspeeder',
passengerCount: '0',
favorited: false },
{ name: 'TIE bomber',
model: 'TIE/sa bomber',
vehicleClass: 'space/planetary bomber',
passengerCount: '0',
favorited: false },
{ name: 'AT-AT',
model: 'All Terrain Armored Transport',
vehicleClass: 'assault walker',
passengerCount: '40',
favorited: false },
{ name: 'AT-ST',
model: 'All Terrain Scout Transport',
vehicleClass: 'walker',
passengerCount: '0',
favorited: false },
{ name: 'Storm IV Twin-Pod cloud car',
model: 'Storm IV Twin-Pod',
vehicleClass: 'repulsorcraft',
passengerCount: '0',
favorited: false },
{ name: 'Sail barge',
model: 'Modified Luxury Sail Barge',
vehicleClass: 'sail barge',
passengerCount: '500',
favorited: false } ]