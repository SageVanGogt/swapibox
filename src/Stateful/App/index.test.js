import React from 'react';
import { shallow } from 'enzyme';
import App from './index';
import films from './../mockData/mockFilms';
import { cleanPeople, cleanPeopleWithValues } from './../mockData/cleanPeople';

describe('App component', () => {
  let mockFilms;
  let mockPeopleData;

  beforeEach(() => {
    mockPeopleData = [{
      homeworld: "Tatooine",
      homeworldPopulation: "200000",
      name: "Luke Skywalker",
      species: "Human",
      favorited: false
    },
    {
      homeworld: "Naboo",
      homeworldPopulation: "4500000000",
      name: "R2-D2",
      species: "Droid",
      favorited: false
    }]
  });

  it('should match the snapshot', () => {
    const wrapper = shallow(<App />, {disableLifecycleMethods: true});        
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render with the correct default state', () => {
    const wrapper = shallow(<App />, {disableLifecycleMethods: true});      
    const expectedAllData = {
      vehicles: [], 
      planets: [], 
      people: [], 
      favorites: [], 
      favoriteCount: 0
    };
    expect(wrapper.state('currentRandomFilm')).toEqual({});
    expect(wrapper.state('currentSectionData')).toEqual([]);
    expect(wrapper.state('allData')).toEqual(expectedAllData);
  });
  
  it('handleFavorite should locate a card in currentSectionData and place it in allData.favorites', () => {
    const wrapper = shallow(<App />, {disableLifecycleMethods: true});    
    const mockEvent = {target: {value: 'Luke Skywalker'}};
    
    wrapper.setState({currentSectionData: mockPeopleData});
    wrapper.instance().handleFavorite(mockEvent);
    
    expect(wrapper.state('allData').favorites.length).toEqual(1);
  });
  
  it('setStateToExistingData should set currentSectionData to data from allData', () => {
    const wrapper = shallow(<App />, {disableLifecycleMethods: true});    
    const allData = {people: mockPeopleData};
    
    wrapper.setState({allData});
    wrapper.instance().setStateToExistingData('people');
    
    expect(wrapper.state('currentSectionData')).toEqual(mockPeopleData);
  });
  
  it('handleClickEvent should call setStateToExistingData if there is data in the allData object', () => {
    const wrapper = shallow(<App />, {disableLifecycleMethods: true});    
    const mockEvent = {target: {name: 'people'}};
    const allData = {people: mockPeopleData};
    const spy =  jest.spyOn(wrapper.instance(), 'setStateToExistingData');
    wrapper.setState({allData});
    wrapper.instance().handleClickEvent(mockEvent);
    
    expect(spy).toHaveBeenCalledWith('people');
  });
  
  it.skip('handleClickEvent should call fetchApiData and setState and allData if the relevant info is not already present', async () => {
    const wrapper = shallow(<App />, {disableLifecycleMethods: true});    
    const mockEvent = {target: {name: 'people'}};
    const mockCleanData = cleanPeople;
    const fetchApiData = jest.fn().mockImplementation(() => (mockCleanData));
    await wrapper.instance().handleClickEvent(mockEvent);
    
  
    expect(wrapper.state('currentSectionData')).toEqual(cleanPeopleWithValues);
  });
  
});
  