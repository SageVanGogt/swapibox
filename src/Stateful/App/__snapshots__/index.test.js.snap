// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="cantina"
          className="cantina-image"
          src="/assets/cantina.jpg"
/>,
        <div
          className="nav-container"
>
          <nav>
                    <Button
                              handleClickEvent={[Function]}
                              section="people"
                    />
                    <Button
                              handleClickEvent={[Function]}
                              section="planets"
                    />
                    <Button
                              handleClickEvent={[Function]}
                              section="vehicles"
                    />
                    <Button
                              favoriteCount={0}
                              handleClickEvent={[Function]}
                              section="favorites"
                    />
          </nav>
          <embed
                    autostart="true"
                    className="music-player"
                    hidden="true"
                    loop="false"
                    src="/assets/Cantina band.mp3"
          />
</div>,
        <Film
          currentRandomFilm={Object {}}
/>,
        <CardContainer
          currentSectionData={Array []}
          handleFavorite={[Function]}
/>,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "cantina",
          "className": "cantina-image",
          "src": "/assets/cantina.jpg",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <nav>
              <Button
                            handleClickEvent={[Function]}
                            section="people"
              />
              <Button
                            handleClickEvent={[Function]}
                            section="planets"
              />
              <Button
                            handleClickEvent={[Function]}
                            section="vehicles"
              />
              <Button
                            favoriteCount={0}
                            handleClickEvent={[Function]}
                            section="favorites"
              />
</nav>,
            <embed
              autostart="true"
              className="music-player"
              hidden="true"
              loop="false"
              src="/assets/Cantina band.mp3"
/>,
          ],
          "className": "nav-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Button
                  handleClickEvent={[Function]}
                  section="people"
/>,
                <Button
                  handleClickEvent={[Function]}
                  section="planets"
/>,
                <Button
                  handleClickEvent={[Function]}
                  section="vehicles"
/>,
                <Button
                  favoriteCount={0}
                  handleClickEvent={[Function]}
                  section="favorites"
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "handleClickEvent": [Function],
                  "section": "people",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "handleClickEvent": [Function],
                  "section": "planets",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "handleClickEvent": [Function],
                  "section": "vehicles",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "favoriteCount": 0,
                  "handleClickEvent": [Function],
                  "section": "favorites",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "nav",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "autostart": "true",
              "className": "music-player",
              "hidden": "true",
              "loop": "false",
              "src": "/assets/Cantina band.mp3",
            },
            "ref": null,
            "rendered": null,
            "type": "embed",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentRandomFilm": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentSectionData": Array [],
          "handleFavorite": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="cantina"
            className="cantina-image"
            src="/assets/cantina.jpg"
/>,
          <div
            className="nav-container"
>
            <nav>
                        <Button
                                    handleClickEvent={[Function]}
                                    section="people"
                        />
                        <Button
                                    handleClickEvent={[Function]}
                                    section="planets"
                        />
                        <Button
                                    handleClickEvent={[Function]}
                                    section="vehicles"
                        />
                        <Button
                                    favoriteCount={0}
                                    handleClickEvent={[Function]}
                                    section="favorites"
                        />
            </nav>
            <embed
                        autostart="true"
                        className="music-player"
                        hidden="true"
                        loop="false"
                        src="/assets/Cantina band.mp3"
            />
</div>,
          <Film
            currentRandomFilm={Object {}}
/>,
          <CardContainer
            currentSectionData={Array []}
            handleFavorite={[Function]}
/>,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "cantina",
            "className": "cantina-image",
            "src": "/assets/cantina.jpg",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <nav>
                <Button
                                handleClickEvent={[Function]}
                                section="people"
                />
                <Button
                                handleClickEvent={[Function]}
                                section="planets"
                />
                <Button
                                handleClickEvent={[Function]}
                                section="vehicles"
                />
                <Button
                                favoriteCount={0}
                                handleClickEvent={[Function]}
                                section="favorites"
                />
</nav>,
              <embed
                autostart="true"
                className="music-player"
                hidden="true"
                loop="false"
                src="/assets/Cantina band.mp3"
/>,
            ],
            "className": "nav-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Button
                    handleClickEvent={[Function]}
                    section="people"
/>,
                  <Button
                    handleClickEvent={[Function]}
                    section="planets"
/>,
                  <Button
                    handleClickEvent={[Function]}
                    section="vehicles"
/>,
                  <Button
                    favoriteCount={0}
                    handleClickEvent={[Function]}
                    section="favorites"
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "handleClickEvent": [Function],
                    "section": "people",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "handleClickEvent": [Function],
                    "section": "planets",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "handleClickEvent": [Function],
                    "section": "vehicles",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "favoriteCount": 0,
                    "handleClickEvent": [Function],
                    "section": "favorites",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "nav",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "autostart": "true",
                "className": "music-player",
                "hidden": "true",
                "loop": "false",
                "src": "/assets/Cantina band.mp3",
              },
              "ref": null,
              "rendered": null,
              "type": "embed",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentRandomFilm": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentSectionData": Array [],
            "handleFavorite": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
